[WooPos][Non Simple Product types] Retry UI for pagination errors on variations screens

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12967 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a new UI for retrying errors during paginating variations.

Along with this, this PR also does a couple of small changes:

In totals screen after payment is success. It navigates the screen back to items list screen so that when we click on "new order" button from the payment success screen, we end up on the items list screen instead of variations screen if we had proceeded to totals screen from variations screen.

Changes the messages which previously only mentioned only simple physical products can be added in POS. After this PR, it should say simple physical and variable products can be added. Below are the places where this got changed:
1. Banner above the items list in items screen
2. On clicking learn more in the banner, a dialog pops up with more information. The changes should reflect here as well
3. When there are no products compatible for POS, we show an empty state. The changes should reflect here as well

#### Design
<img width="458" alt="Screenshot 2024-11-19 at 7 19 03 AM" src="https://github.com/user-attachments/assets/674a6a5d-701f-437c-b0ee-e70eda5d6b72">

#### Implementation
<img width="379" alt="Screenshot 2024-11-21 at 1 42 06 PM" src="https://github.com/user-attachments/assets/2eb97aa1-92ac-42f9-9dbd-98e9eba17ff8">


### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
### Retry pagination UI
1. Navigate to POS mode (more menu -> POS)
4. Click on any variable products that has more than 10 variations
5. Ensure you navigate to the variations screen
6. Turn on airplane mode
7. Scroll down
8. Ensure you see the pagination error UI
9. Click on retry and ensure nothing happens as we are still in airplane mode
10. Turn off airplane mode
11. Now, click on retry button again
12. Ensure the variations gets loaded

### Navigate to items list after payment success
1. Navigate to POS mode (more menu -> POS)
2. Add products to cart
3. **Navigate to variations screen before checking out**
4. Ensure payment is success
5. Click on "New Order" button
6. Ensure you navigate to items list screen.

### Changes to messages
1. Navigate to POS mode (more menu -> POS)
2. Ensure you see a banner that says both simple and variable products instead of just simple products
3. Click on learn more in the banner
4. Ensure the dialog also mentions both simple and variable products


1. Navigate to POS mode where there are no compatible products to display (no simple and variable products)
2. Ensure the empty state also says that only simple and variable products are compatible.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Tested on both light and dark modes

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/20818862-21db-4ee2-b69f-5d25d5923162

<img width="634" alt="Screenshot 2024-11-22 at 9 35 47 AM" src="https://github.com/user-attachments/assets/9dee7305-83bb-4978-a278-c86549b9a879">
<img width="803" alt="Screenshot 2024-11-22 at 9 35 55 AM" src="https://github.com/user-attachments/assets/49ac4d38-415c-4e83-a8a1-acdd07ed982c">




- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Add state to indicate error while paginating

AnirudhBhat committed:
Handle error while paginating in WooPosItemsList.kt

AnirudhBhat committed:
Handle error while paginating in WooPosItemsScreen.kt

AnirudhBhat committed:
Remove snack bar implementation from WooPosVariationsScreen

AnirudhBhat committed:
Override errorloadingMoreItems property in variations view state

AnirudhBhat committed:
Navigate to items list screen on payment success so that clicking on new order button will take us to items list screen instead of variations screen if the user proceeded to checkout while they were in variations screen.

AnirudhBhat committed:
Add tests for different state updates while paginating

AnirudhBhat committed:
Remove unnecessary padding from the variations screen

AnirudhBhat committed:
Add test to verify we navigate to the items list screen on payment success

AnirudhBhat committed:
Remove unused imports

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>VariationsLRUCache</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kidinov commented:
@AnirudhBhat 👋

Where does this design come from? IMO, it doesn't match expectations from this kind of UX in a mobile app.

I believe this should have:

* The same height as a normal item
* Just one line: icon | text | button
* When the button is clicked, we replace that with the "loading indicator" list item till the response comes

Something like that maybe?

<img width="807" alt="image" src="https://github.com/user-attachments/assets/dfdfe1dd-4049-4eb0-8cf1-792e0d330b8a">


@malinajirka @samiuelson wdyt?

AnirudhBhat commented:
>Where does this design come from? IMO, it doesn't match expectations from this kind of UX in a mobile app.

>I believe this should have:

>The same height as a normal item
>Just one line: icon | text | button
>When the button is clicked, we replace that with the "loading indicator" list item till the response comes
>Something like that maybe?

Wagner gave the designs and we had conversation about this on Slack [here](https://a8c.slack.com/archives/C070SJRA8DP/p1731901583751489)  

IMHO, I don't see any problem with this design and I don't feel it deviates from the mobile UX especially considering that this is run only on tablet devices.

1. Clarity: It provides a clear, standalone visual for errors, which ensures users can immediately understand what went wrong.
2. Consistency: It aligns with the rest of our app's (POS) error screens, it maintains uniformity and helps users understand that something went wrong quickly.
3. Prominent Call to Action: The "Retry" button is clearly visible and easy to tap - as this was something we considered for POS that CTAs should be big and easily tappable considering our use case is for the POS.

If we show the error as another card with the same height as rest of the items. I can see few confusions that could arise:
1. Clarity: On a quick glance, as the user will be scrolling, it's hard to understand what happened. Whether a new item got loaded or was it an error.
2. Inconsistency: The way we show the error here is inconsistent with the rest of the screens on our POS system.
3. Confusion: The card-style error display, with the same height as a list item, might create confusion by making it appear as though only one item failed to load, rather than multiple items or the entire batch.

What do you think?

AnirudhBhat committed:
Make changes to string resource to include variable products as well where it said we support only simple physical products.

malinajirka commented:
I lean towards the approach Andrei mentioned (mostly because it's a standard behavior across apps on Android), but it's ultimately Wagner's decision, so I'd defer this to him. 

One thing which feels broken is the animation that happens after you tap on `Retry` in the video - I'm not even sure, if more items loaded or if just the scrolling position changed. If we decide to go with this current approach, I think we need to polish it. If we decide to go with Andrei's approach, this problem disappears as the item will be replaced with the existing shimmer effect row so there won't be any change in the scrolling position at all.

kidinov commented:
@AnirudhBhat 👋

> Clarity: It provides a clear, standalone visual for errors, which ensures users can immediately understand what went wrong.

With the item being so big, I have doubts that people will even understand what's going on, as your synthetical example is not what people will have. They will have a list of 20 or more products. They will scroll down where they will see an item with a shimmer effect, which after some time is replaced by a huge thing, but that huge thing won't be visible automatically. They will see, in most cases, just empty space at the bottom People will need to figure out that they need to scroll down more to find this "retry" button, and there is no necessity at all to have that thing so huge and so not in line with the rest of the items. On top of that the animation looks really wrong when things are that huge.

And again, this is completely not what people got used to see in mobile apps, so clarity is really questionable here

> Consistency: It aligns with the rest of our app's (POS) error screens, it maintains uniformity and helps users understand that something went wrong quickly.

But this is not an error screen at all. It's a contextual error. People can easily keep using the shown products if the next page is not loaded

> Prominent Call to Action: The "Retry" button is clearly visible and easy to tap - as this was something we considered for POS that CTAs should be big and easily tappable considering our use case is for the POS.

The button in the item can be as huge as we want/need



Looking into Figma design, I feel that there is a misunderstanding on the flow. In 2 out of 3 designs, there are no products shown at all. I'll ask Wagner to reconsider

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosBaseViewState.kt:
```diff
@@ -8,4 +8,5 @@ interface ContentViewState {
     val items: List<WooPosItem>
     val loadingMore: Boolean
     val reloadingProductsWithPullToRefresh: Boolean
+    val errorLoadingMoreItems: Boolean
```
The state defined like that can contain invalidate states, e.g. `loadingMore = true` and `errorLoadingMoreItems = true`. 

We probably should do something like
```
val paginationIndicator: PaginationIndicator

enum PaginationIndicator {
   Loading, Error, None
} 
```

By the way, `ContentViewState` does not follow the naming convention of the top-level classes. It probably should be `WooPosItemsViewState.` And there are a few more, e.g., `ToolbarAccessibilityLabels`, for instance. And some classes that are in the `items` package now, e.g., `WooPosBanner` should either contain an item in them or be placed in another package

AnirudhBhat replied:
Thanks for raising this. I had this on my list to refactor hence I kept the PR as draft. Here's the refactored state: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/a45bd6cfc7e2eb42c0cf0916f80f98db76ef07e0

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsScreen.kt:
```diff
@@ -118,93 +92,82 @@ private fun WooPosVariationsScreens(
+            .padding(
+                start = 16.dp.toAdaptivePadding(),
+                end = 16.dp.toAdaptivePadding(),
+                top = 30.dp.toAdaptivePadding(),
```
I think 30 can not be in 4dp grid. 28 or 32 should be

https://m2.material.io/design/layout/spacing-methods.html#baseline-grid

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/71ee602d411b0cba2692198a59e3c08db12f37c3

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -100,18 +93,15 @@ class WooPosVariationsViewModel @Inject constructor(
         loadMoreJob = viewModelScope.launch {
             val result = variationsDataSource.loadMore(productId)
             if (result.isSuccess) {
                 Result.success(Unit)
```
What does this line do?

AnirudhBhat replied:
This is no longer relevant as result of this change: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/a45bd6cfc7e2eb42c0cf0916f80f98db76ef07e0

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -100,18 +93,15 @@ class WooPosVariationsViewModel @Inject constructor(
         loadMoreJob?.cancel()
         loadMoreJob = viewModelScope.launch {
             val result = variationsDataSource.loadMore(productId)
             if (result.isSuccess) {
```
np:
```
            _viewState.value = if (result.isSuccess) {
                currentState.copy(loadingMore = false, errorLoadingMoreItems = false)
            } else {
                currentState.copy(loadingMore = false, errorLoadingMoreItems = true)
            }
```

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -100,18 +93,15 @@ class WooPosVariationsViewModel @Inject constructor(
             if (result.isSuccess) {
                 Result.success(Unit)
-                if (!variationsDataSource.canLoadMore()) {
-                    _viewState.value = currentState.copy(loadingMore = false)
```
Can you use currentState here?  `_viewState.value` can be modified by the time the request comes back but you copying based on the old state

AnirudhBhat replied:
This is no longer relevant as result of this change: https://github.com/woocommerce/woocommerce-android/commit/a45bd6cfc7e2eb42c0cf0916f80f98db76ef07e0

AnirudhBhat committed:
Refactor pagination state logic. Use sealed interface instead of boolean value to minimise the occurence of invalid states

AnirudhBhat committed:
Yse 32 instead of 30 dp.

AnirudhBhat committed:
Use MutableStateFlow's atomic operation to update the view state

AnirudhBhat committed:
fix detekt error

AnirudhBhat committed:
Add string resource for pagination error messages

AnirudhBhat committed:
Add pagination error component for items list

AnirudhBhat committed:
Integrate the pagination error component in items and variations screens

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat commented:
@kidinov Thanks for the review. I've made the changes so that it shows retry pagination UI as a card item now. Let me know what you think of it


https://github.com/user-attachments/assets/5532d310-9b75-42fa-8c15-f8d0585c3428

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosPaginationErrorScreen.kt:
```diff
@@ -0,0 +1,111 @@
+package com.woocommerce.android.ui.woopos.common.composeui.component
```
Just noticed that in many places "toAdaptivePaddings" are not used, so the layout will look bad on different screen types

AnirudhBhat replied:
Fixed. Let me know if I have missed anywhere else - https://github.com/woocommerce/woocommerce-android/pull/12972/commits/e94d0e5c2def32ecae636e7aea39d72994b37bb1

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosBaseViewState.kt:
```diff
@@ -6,6 +6,12 @@ sealed class WooPosBaseViewState(
+    val paginationState: PaginationState
+}
+
+sealed interface PaginationState {
```
Just curious, why interface?

AnirudhBhat replied:
Just for flexibility as this paginationState will be used for different kind of items in future and I wasn't sure if i want to tie up the subtype into a restricted base class. 

At this point. It doesn't make any difference whether it is a sealed class or interface.

kidinov replied:
> At this point. It doesn't make any difference whether it is a sealed class or interface.

Imo sealed interface has to be used on purpose, and if there is no purpose, then it's better to use class to avoid confusion

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/a9780a3262ed80eb9e995dbbb1dd58c1e0ca432c

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosItemsScreen.kt:
```diff
@@ -169,15 +170,21 @@ private fun MainItemsList(
                 is WooPosItemsViewState.Empty -> ProductsEmptyList()
 
-                is WooPosItemsViewState.Error -> ProductsError { onRetryClicked() }
+                is WooPosItemsViewState.Error -> ProductsError(modifier = Modifier.width(640.dp)) { onRetryClicked() }
```
Where does 640 come from? How does this look on a screen of 800 or whatever min we support now?

AnirudhBhat replied:
This code was already present in the `trunk`. I just re-arranged it.

kidinov replied:
it's a bug I beleave, we should fix that probably not expand it to another screen

AnirudhBhat replied:
Just to understand more. What problem do you see using width of 640? We could use fillMaxWidth() instead.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -93,25 +88,36 @@ class WooPosVariationsViewModel @Inject constructor(
-                    _viewState.value = currentState.copy(loadingMore = false)
+        viewModelScope.launch {
+            val updatedState = _viewState.updateAndGet { currentState ->
+                if (currentState is WooPosVariationsViewState.Content && variationsDataSource.canLoadMore()) {
```
Could you please explain why those extra complications are needed? There is the method which does the same in the products already and it looks much shorter and clear

```
    private fun onEndOfProductsListReached() {
        val currentState = _viewState.value
        if (currentState !is WooPosItemsViewState.Content) {
            return
        }

        if (!productsDataSource.hasMorePages) {
            return
        }

        _viewState.value = currentState.copy(paginationState = PaginationState.Loading)

        loadMoreProductsJob?.cancel()
        loadMoreProductsJob = viewModelScope.launch {
            val result = productsDataSource.loadMore()
            _viewState.value = if (result.isSuccess) {
                result.getOrThrow().toContentState()
            } else {
                WooPosItemsViewState.Error()
            }
        }
    }
```

AnirudhBhat replied:
As you mentioned, we rely on `currentState` here instead of `_viewState` and this means there is a possibility that we might not be dealing with the latest view state. In order to avoid this, I used thread safe methods of mutable flow.

I've further refactored for readability. Let me know what you think

https://github.com/woocommerce/woocommerce-android/pull/12972/commits/6fe46f232ea601f3d12f10dca8bd4611344b60b0

kidinov replied:
> this means there is a possibility that we might not be dealing with the latest view state
Sorry, I am not following. There is no suspension or anything, and it's all on the main thread. When we launch a new courutine we don't use `currentState` anymore

kidinov reviewed (commented):
@AnirudhBhat not for this PR probably but, I would suggest looking into the details of implementation of `WooPosProductsDataSource` and mimick the same with variations:
* For consistency
* There are actually issues with the current implementation; for instance, `canLoadMore` is a var that changed from different threads. Or that you create a new flow every time you call `getVariationsFlow`

AnirudhBhat committed:
Use adaptive padding

AnirudhBhat committed:
Move the modifier parameter near WooPosErrorScreen

AnirudhBhat committed:
Update the state to pagination error on items view model

AnirudhBhat committed:
Refactor loadMore method for readability

AnirudhBhat committed:
Refactor variation flow collection logic

AnirudhBhat commented:
@kidinov Thanks for the review. I've addressed the feedback and have refactored the [variation collection logic](https://github.com/woocommerce/woocommerce-android/pull/12972/commits/0f5cf15ab93d54ef47eea5ffc9ff32b9434e092a) (Thanks for catching this 🙇 ). Please take a look. 

Since Wagner mentioned in our internal Slack group that he will be providing a new UI for the pagination error. I haven't focused much on the UI changes here.

AnirudhBhat committed:
Merge branch 'issue/12846-analytics-for-variations' into issue/12967-retry-pagination-error

AnirudhBhat committed:
Fix failing test

AnirudhBhat committed:
Merge branch 'trunk' into issue/12967-retry-pagination-error

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosPaginationErrorScreen.kt:
```diff
@@ -0,0 +1,112 @@
+import com.woocommerce.android.ui.woopos.common.composeui.toAdaptivePadding
+
+@Composable
+fun WooPosPaginationErrorScreen(
```
np: maybe call it somehow closer to reality? `WooPosPaginationErrorIndicator`?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/aa38f102d6d4553308fce6a009a3cd48ea14c297

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosItemsScreen.kt:
```diff
@@ -317,6 +324,17 @@ fun ProductsError(onRetryClicked: () -> Unit) {
 }
 
+@Composable
+fun ProductsPaginationError(onRetryClicked: () -> Unit) {
```
if it's public then it should be called with `WooPos...`

kidinov replied:
And in other places the same

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/6a3b7772a5188a9b40690b7f98a73cadec6c3e96

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosItemsScreen.kt:
```diff
@@ -353,7 +371,7 @@ fun WooPosItemsScreenPreview(modifier: Modifier = Modifier) {
                 ),
             ),
-            loadingMore = true,
+            paginationState = PaginationState.Loading,
```
Shell we also have preview for the pagination error?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/6bb0870382adeedb9777e488f7b21c0194e5c4c4

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosPaginationErrorScreen.kt:
```diff
@@ -0,0 +1,112 @@
+        modifier = modifier
+            .fillMaxWidth()
+            .height(112.dp)
+            .clip(RoundedCornerShape(16.dp))
```
If the goal was to mimic the product's items, then it doesn't match. Please take a look at `WooPosItemsList` and copy paste the card and row setup from there

Btw,  in `WooPosItemsList` there is naming without `WooPos` of a public `ItemList`

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/19a7ed5aa5f6b35c1cb71e84d6a6a60511292b36

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosPaginationErrorScreen.kt:
```diff
@@ -0,0 +1,112 @@
+        verticalAlignment = Alignment.CenterVertically,
+        horizontalArrangement = Arrangement.SpaceBetween
+    ) {
+        Row(
```
Why do you need 3 rows nested here

AnirudhBhat replied:
Refactored here: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/19a7ed5aa5f6b35c1cb71e84d6a6a60511292b36

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosItemsScreen.kt:
```diff
@@ -317,6 +324,17 @@ fun ProductsError(onRetryClicked: () -> Unit) {
+    WooPosPaginationErrorScreen(
+        message = stringResource(id = R.string.woopos_items_pagination_error),
+        primaryButton = Button(
+            text = stringResource(id = R.string.woopos_items_pagination_load_more_label),
```
Not sure if it's missed but in both places it's still `Failed to load more items. Please try again.` with `Load more`, while in preview 

```
    WooPosTheme {
        WooPosPaginationErrorScreen(
            message = "Error loading products",
            primaryButton = Button(
                text = "Load more",
                click = {}
            )
        )
    }
```

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/623125f8a7bfb9e4a38ff6df96c93bbf7a750f14

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosPaginationErrorScreen.kt:
```diff
@@ -0,0 +1,112 @@
+@WooPosPreview
+fun WooPosPaginationErrorScreenPreview() {
+    WooPosTheme {
+        WooPosPaginationErrorScreen(
```
I am considering whether a preview with a list or a list with outer padding would be better for development and debugging.

AnirudhBhat replied:
We already have a preview like this in `WooPosItemsScreen`. Let me know if I have misunderstood your query.

kidinov replied:
I find it a matter of personal preference, but switching between `WooPosItemsScreen` and the code in this file while changing card settings is not the most enjoyable experience. On top of that it takes time to rerender `WooPosItemsScreen` as it's huge and has multiple reviews

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/1eb6c32d9189d57965ba2bbb47919fb2185dee5b

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsScreen.kt:
```diff
@@ -118,93 +93,82 @@ private fun WooPosVariationsScreens(
 
-                    else -> {}
+                is WooPosVariationsViewState.Error -> {
+                    VariationsError(modifier = Modifier.width(640.dp)) {
```
What is the reason for that? We are showing empty space now

<img width="388" alt="image" src="https://github.com/user-attachments/assets/4abc5886-592d-480d-ade5-7351e98257b0">

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsScreen.kt:
```diff
@@ -118,93 +93,82 @@ private fun WooPosVariationsScreens(
-                        }
-                    }
+                is WooPosVariationsViewState.Loading -> ItemsLoadingIndicator(
+                    minOf(10, variableProductData.numOfVariations)
```
If the goal is to show the exact amount of variations as loading items, then I don't understand why we need to take min from it or 10?

AnirudhBhat replied:
The goal is not to show the exact amount of loading items as variations. The goal is to show the exact amount of loading items as variations only if the variation is <= 10.

kidinov replied:
Then, I don't understand the logic behind that. I think we either want to indicate the number of items we load or don't want to do it and just show a generic loading screen similar to the products. Imo this just creates confusion for both us and the users on top of that it leaks logic to the view layer

AnirudhBhat replied:
Alright, fixed here: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/8c3af251730a505d20a2ec19f903528547365cc4

kidinov replied:
That's more a question to Wagner, but to me, if we know beforehand how many items we are loading and as we even show the number at the top, show 10 loading items when we load 1 or 5 is a bit off

AnirudhBhat replied:
I'll ask Wagner what he feels here.

kidinov commented:
https://github.com/user-attachments/assets/c1bbd1cb-5a34-4784-97c4-2457638e769f

https://github.com/user-attachments/assets/28bc9a21-aaec-4ade-b9ea-1c85466d3cb2


@AnirudhBhat 

some things I noticed:
* 7 variations -> During loading still shows one loading indicator
* 4 variations -> shows only 1 and every time you open it tries to load another page why clearly there is no
* It actually loads something. When I simulate error it can end up with pagination error UI which seems to be doing another on click

```
    suspend fun loadMore(productId: Long): Result<Unit> = withContext(Dispatchers.IO) {
        if (Math.random() > 0.5) {
            delay(3000)
            Result.failure(Exception("Random error"))
        } else {
            val result = handler.loadMore(productId)
            if (result.isSuccess) {
                Result.success(Unit)
            } else {
                result.logFailure()
                Result.failure(
                    result.exceptionOrNull() ?: Exception("Unknown error while loading more variations")
                )
            }
        }
    }
```

AnirudhBhat committed:
Fix failing test

AnirudhBhat committed:
Rename WooPosPaginationErrorScreen.kt to WooPosPaginationErrorIndicator

AnirudhBhat committed:
Make the method private

AnirudhBhat committed:
Add preview for pagination error state

AnirudhBhat committed:
Refactor WooPosPaginationErrorIndicator and add content description for it.

AnirudhBhat committed:
Rename ItemList to WooPosItemList

AnirudhBhat committed:
Use string resource in preview

AnirudhBhat committed:
Use sealed class instead of interface for pagination state

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Revert change

AnirudhBhat committed:
Default canLoadMore value to false

AnirudhBhat committed:
Remove minOf logic

AnirudhBhat committed:
Add list for WooPosPagina' preview

AnirudhBhat committed:
Add margin for items loading in variations screen

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -93,29 +100,43 @@ class WooPosVariationsViewModel @Inject constructor(
-        val currentState = _viewState.value
-        if (currentState !is WooPosVariationsViewState.Content) {
-            return
+        val updatedState = _viewState.updateAndGet { currentState ->
```
Could you please explain why the approach taken in the simple products won't work here? It just feels so much extra code and complexity not clear what for 

```
    private fun onEndOfProductsListReached() {
        val currentState = _viewState.value
        if (currentState !is WooPosItemsViewState.Content) {
            return
        }

        if (!productsDataSource.hasMorePages) {
            return
        }

        _viewState.value = currentState.copy(paginationState = PaginationState.Loading)

        loadMoreProductsJob?.cancel()
        loadMoreProductsJob = viewModelScope.launch {
            val result = productsDataSource.loadMore()
            _viewState.value = if (result.isSuccess) {
                result.getOrThrow().toContentState()
            } else {
                currentState.copy(paginationState = PaginationState.Error)
            }
        }
    }
```

AnirudhBhat replied:
Simplified the logic: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/9dfffca62c5c42f6421d61f44d42d68e3f5d81ad

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -93,29 +100,43 @@ class WooPosVariationsViewModel @Inject constructor(
         }
 
     fun loadMore(productId: Long) {
```
Its public on purpose? If this is for testing then it can be replaced with just push `EndOfItemsListReached` event to simulate the needed situation

AnirudhBhat replied:
This is not intended, fixed here: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/780a738654d3a245d2f124ef92a9a3458d331198

kidinov commented:
@AnirudhBhat 👋

There is no loading state when opening variations for the first time now? Reproducible when variation opened for the first time


https://github.com/user-attachments/assets/ae88274b-e3d4-4e40-b6a2-925fe3a4c417


There is still something with pagination. Notice that items jump on variations and do not jump on simple products.

https://github.com/user-attachments/assets/0151f560-0ca0-4e04-9c41-199c6a94ae89

Full-screen error state also doesn't work properly. It's shown, and then we just push locally stored variations.

https://github.com/user-attachments/assets/7cbf445c-ac49-4dc4-9453-d684d22c9bfc

AnirudhBhat committed:
Use drop to discard the cache value

AnirudhBhat committed:
Make the test pass

AnirudhBhat committed:
Make the method private

AnirudhBhat committed:
Fix jumping of items while paginating

AnirudhBhat committed:
Simplify load more logic

AnirudhBhat commented:
@kidinov Thanks for all the reviews. I've fixed the issues raised. Please have a look.

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/12967-retry-pagination-error

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModelTest.kt

AnirudhBhat committed:
Make the test pass

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/12967-retry-pagination-error

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModelTest.kt

kidinov committed:
Merge branch 'trunk' into issue/12967-retry-pagination-error

AnirudhBhat committed:
Fix import ordering.

kidinov commented:
@AnirudhBhat Thanks for the improvement. Looks better now!

A few findings from my side

It looks like we are loading the variations every time, even when they should be cached already? We indeed should refetch them, but we need to show the cache right away and then update the list with newly fetched data, the same as on the products list

https://github.com/user-attachments/assets/70bc19ae-d43e-4cde-bd5c-efdde6a6dba4

For some reason, the bottom padding disappeared for the products and variations too. They were there on purpose because of the connection buttons and menu. Now the error text of the pagination item is not readable at all


https://github.com/user-attachments/assets/6a505dc3-482c-4216-99de-cb2cbcf4d085

It's less now, but the items are still jumping when the next page is loaded. It's not a big problem itself, but it means that some items are recreated which usually indicates some identification issue

https://github.com/user-attachments/assets/f7790b3e-89b1-4baa-b7ee-f3e772c00161

AnirudhBhat committed:
Cache variations and fix variations jumping while paginating

AnirudhBhat committed:
Make tests pass

AnirudhBhat committed:
Remove unused import

AnirudhBhat commented:
@kidinov I've added caching and fixed other 2 issues. Please have a look.

kidinov commented:
@AnirudhBhat 

It feels like something was fixed but something got broken =)

On a general note, in my opinion, the idea that we subscribe to the local data store changes and show only them can not work well. You have competing sources of truth -> errors from your "hooks" and asynchronously pushed data, which we never know when it arrives. I think it's impossible to make it work properly with this approach

Now, the error screen is never shown on the variations. I'd suggest you look into `WooPosProductsDataSource.` I believe  the way how both `VariationListHandler` and `ProductListHandler` exposes their APIs is not possible to implement properly the lists for our requirements. That's why we adapted it in `WooPosProductsDataSource` with it's own in memory cache and separated remote and cached data. 

https://github.com/user-attachments/assets/740790a4-27b1-4e6c-9b61-ab700a5636e9

---

The bottom padding is still not the same as it used to be; part o the item is covered.

<img width="894" alt="image" src="https://github.com/user-attachments/assets/8bb085af-3238-4eb2-a467-f0da68bfcd0c">

---

And the bottom padding when it's the end of the list seems enormous now. I believe it was much smaller

<img width="976" alt="image" src="https://github.com/user-attachments/assets/6a81975c-025b-4ec9-930d-dff173fd15d6">

---

My memory is a bit blurry, and I can not test right now as the error state doesn't work. But I think there  a reason why we clean and hide the card when there is a fullscreen error on the products. To avoid the situation when a product not in the cache but an order is created with it e.g.:
* Open variations
* Add one
* Do pull to refresh -> error happens
* Card will be still visible and the variation I believe will be there too
* You can still even go to checkout -> I believe nothing good will happen, maybe even a crash

---

I got this again on a real device. To reproduce just generate more products on the backend and do pull to refresh. I didn't check, but I'd suggest looking into what is the key for variable products Is it actually unique or we need some composite key for them?

https://github.com/user-attachments/assets/e6ae6045-1558-4c5e-98fa-24735fea4dc4

---

I think I mentioned that already. We show for instance "2 variations", but at the same time there are 10 or something like that loading items. Not a big deal but as I remember you already passed the numbers of variations anyway, seem worth just passing this value to indicate how many variations we are loading, wdyt?

https://github.com/user-attachments/assets/bbba611f-7fd2-489e-8434-6d1faab530b9

--- 

Even when the variations are in the cache, there is still this weird blinking. It's not a big deal, but considering that we don't have this for products, why not just copy exactly the same from there?

And the same happens for some reason for the next page loading. On the products, it appears right away while on variations, there is a very visible lag

https://github.com/user-attachments/assets/9920b1e6-af6c-44cf-85b2-3b6de874b68e

---

Pagination still works weirdly. Notice that at the end of the video, we were loading something, but nothing was loaded


https://github.com/user-attachments/assets/4cfd2459-a294-44c8-9336-a087b94ba780

AnirudhBhat committed:
Use in memory cache

AnirudhBhat committed:
Remove unnecessary code

AnirudhBhat commented:
@kidinov Ah, that's frustrating. I've used in-memory cache now and hopefully, it should fix all the issues you mentioned above. Please have a look.

>I think I mentioned that already. We show for instance "2 variations", but at the same time there are 10 or something like that loading items. Not a big deal but as I remember that you already pass the numbers of variations anyway, seems worth improvement, wdyt?

If rest of the issues are fixed, I'll take this in a separate PR. Let me know.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/VariationsLRUCache.kt:
```diff
@@ -0,0 +1,48 @@
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+class VariationsLRUCache<K, V>(private val maxSize: Int) {
```
* I would be very surprised if the Android SDK or Kotlin SDK doesn't have implementation of the LRU cache
* Why do we even need an LRU cache? I think simple LinkedHashMap should do just fine -> just remove the first inserted from the list when it gets bigger than you want

AnirudhBhat replied:
I've used the in-built implementation of LRU cache now. Thanks for pointing this out. 

Regarding LinkedHashMap, Yes we could use it but we need to manually write the eviction logic as you mentioned. I don't see any harm or maintainance problem with LRU cache as well.

kidinov replied:
@AnirudhBhat check the naming of this class

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt:
```diff
@@ -13,41 +18,77 @@ import javax.inject.Singleton
-    fun getVariationsFlow(productId: Long): Flow<List<ProductVariation>> {
-        return handler.getVariationsFlow(productId)
+    private val variationCache = VariationsLRUCache<Long, List<ProductVariation>>(maxSize = 50)
+    private val cacheMutex = kotlinx.coroutines.sync.Mutex()
```
np: should we use import?

AnirudhBhat replied:
Done: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/7e17092c29fbf56426f0ee0d52abe3c8b2e3a44a

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -37,50 +37,89 @@ class WooPosVariationsViewModel @Inject constructor(
-    val events = _events.asSharedFlow()
-
-    private var fetchJob: Job? = null
+    private var flowJob: Job? = null
```
Never used?

AnirudhBhat replied:
Made some changes here: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/a61e5f65609f1e19685be0354aca115a5653ba54

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -37,50 +37,89 @@ class WooPosVariationsViewModel @Inject constructor(
 
-    fun init(productId: Long) {
-        fetchVariations(productId = productId, withPullToRefresh = false, withCart = true)
+    fun init(productId: Long, withPullToRefresh: Boolean = false, withCart: Boolean = true) {
```
There are default params that you never pass to this method

AnirudhBhat replied:
Removed default parameters here: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/b23cd264697f32d31038b7b9cf188a0aeb5002bc

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -37,50 +37,89 @@ class WooPosVariationsViewModel @Inject constructor(
 
-    fun init(productId: Long) {
-        fetchVariations(productId = productId, withPullToRefresh = false, withCart = true)
+    fun init(productId: Long, withPullToRefresh: Boolean = false, withCart: Boolean = true) {
```
Overall, have you explored the possibility of recreating this VM every time the variations are open? Now it seems very error prone as the state is kept for different lists of variations

AnirudhBhat replied:
I had and navigating to and from variation screen was causing problem. Problem in the sense that we need only left pane to change when coming back from variations screen but it was looking as if the whole screen was being popped out of the back stack. I've slightly modified and the View Model is recreated now for each variable product: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/1718bb33d8b007b3e5d7defb2b2d0e9437adef9d

kidinov commented:
@AnirudhBhat 👋

Looks much better!

You show old data for a second before it gets updated now


https://github.com/user-attachments/assets/a3b65997-926f-4cdd-ae5e-3e03da902708

---

Sometimes it loads the next page but loads nothing:

https://github.com/user-attachments/assets/37ec9fb6-61ee-42c9-a090-3eb40726f0d9

---

Why page size is 10 items for variations, while for products 25? I know that you reuse the handler, but I am not sure that it should be different in our POS case

---

We won't load the next page if the response hasn't come back yet. It's tricky, but:
* having 25 items page size would partially mitigate that
* Or you actually know the exact size of the variation list, you can easily determine when the next page exists and when not without getting response

https://github.com/user-attachments/assets/2f8a1088-34e5-4ff9-bdf9-826850997cdd

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/selector/VariationListHandler.kt:
```diff
@@ -42,4 +41,8 @@ class VariationListHandler @Inject constructor(private val repository: Variation
     }
+
+    fun resetLoadMoreState() {
+        canLoadMore = false
```
It looks like you modified the atomic var boolean from a different thread, which may lead to inconsistency. Check `ProductListHandler`

Having said that you may need determine if there is a next page based on specific products. Now it's shared state between different lists which may or already leads to issues

AnirudhBhat replied:
This is no longer needed. I've removed this method.

AnirudhBhat committed:
Use SDK's in built LRU cache

AnirudhBhat committed:
Update variations page size to 25

AnirudhBhat committed:
Remove unnecessary method

AnirudhBhat committed:
Add key to hilt view model so that new instance is created for each key

AnirudhBhat committed:
Remove unnecessary methods

AnirudhBhat committed:
Remove unnecessary val

AnirudhBhat committed:
Use import for Mutex

AnirudhBhat committed:
Remove default parameters

AnirudhBhat commented:
@kidinov Thanks for your patience and the reviews. Much appreciated 🙇 

>You show old data for a second before it gets updated now

I couldn't reproduce this to be honest. This should be fixed now with view model being re-created for each variable product

>Sometimes it loads the next page but loads nothing:

Can you please let me know the exact steps? I couldn't reproduce this either.

>We won't load the next page if the response hasn't come back yet. It's tricky

I've set page size to 25 now and it should to some extent fix the issue. I'll come up with a separate PR to see how best to avoid this situation. I hope that's okay with you.

Also, can you please do another round of testing. Thanks!

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/VariationsLRUCache.kt:
```diff
@@ -0,0 +1,41 @@
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+class VariationsLRUCache<K, V>(maxSize: Int) {
```
It looks like that most of the methods of this class are not used

AnirudhBhat replied:
Removed unnecessary methods: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/f1f54b0471ceeae3299350250c7519d240474a65

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsDataSource.kt:
```diff
@@ -13,30 +19,57 @@ import javax.inject.Singleton
+    private val cacheMutex = Mutex()
+
+    private suspend fun getCachedVariations(productId: Long): List<ProductVariation> {
+        return cacheMutex.withLock { variationCache.get(productId) ?: emptyList() }
```
You have mutex here and you have another mutex on mutex inside of the cache

AnirudhBhat replied:
Remove unnecessary mutex locks: https://github.com/woocommerce/woocommerce-android/pull/12972/commits/dfaa6fea62c2e4cc97eabd89ea276ca8cd5998a1

kidinov commented:
@AnirudhBhat 👋

It seems mostly fine now. I left a couple of code-related comments, though

And I still have this very weird behavior with the pagination of the variations. Notice that the loading item disappears earlier sometimes, sometimes it doesn't appear at all . I'd suggest you to use https://github.com/woocommerce/wc-smooth-generator plugin and try different cases yourself with different amounts of variations to fix the issue

https://github.com/user-attachments/assets/080cde22-43de-4963-9816-9371e4cd1e3b

AnirudhBhat committed:
Remove unnecessary methods

AnirudhBhat committed:
Remove unnecessary mutex lock

AnirudhBhat committed:
Use load more job to show pagination state

AnirudhBhat committed:
Add test to verify we show pagination state loading when updating view state if load more job is active

AnirudhBhat committed:
Add test to verify we show pagination state None when updating view state if load more job is not active

AnirudhBhat committed:
Fix detekt error

AnirudhBhat committed:
Handle pagination state for remote fetch

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/12967-retry-pagination-error

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsScreen.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModelTest.kt

AnirudhBhat commented:
@kidinov I was able to figure out the pagination edge case issue. Here's my understanding of the problem:

The issue occurs because when I open the variations screen, cached data is displayed first, and a background remote fetch is triggered to refresh the data. If I scroll to the bottom during this time, a loadMore operation is triggered to fetch the next page of variations, and the pagination loading state is displayed. However, when the initial remote fetch completes, it updates the view state and unintentionally resets the pagination loading state, making it look like nothing is happening. This is misleading because the loadMore operation is still in progress, and the user has to wait without any indication until the additional data is finally loaded. 

This was reproducible on the products list as well.

https://github.com/user-attachments/assets/66ce25a3-527f-485e-bae6-6996f5cbd961


I've made the changes and it should fix the problem. Please have a look.

kidinov commented:
@AnirudhBhat 

I still consistently can reproduce the issue 

https://github.com/user-attachments/assets/8962c8ff-70f7-4d8a-9b1e-242d503231f8

AnirudhBhat committed:
Add additional condition to determine whether to load more or not in the variations screen.

AnirudhBhat commented:
@kidinov Can you please have a look now. Let me know if it's still reproducible and I'll probably create a new ticket for it and unblock this PR. Many thanks!

kidinov commented:
@AnirudhBhat I could not able to reproduce the issue anymore

We'll need to implement the loading more error in this PR, though, as it is also used in products, right?

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/12967-retry-pagination-error

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt
#	WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModelTest.kt

AnirudhBhat commented:
>We'll need to implement the loading more error in this PR, though, as it is also used in products, right?

I'll implement the new updated designs in a separate PR. I want to unblock this PR as it has been open for quite some time.

kidinov reviewed (approved):
@AnirudhBhat 👍

Just remember that the loading more UI won't be merged before the cut off then the mock UI will be visible to the people

AnirudhBhat committed:
Merge branch 'refs/heads/trunk' into issue/12967-retry-pagination-error

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt
